packages/render/src/use-fetch-frame.ts

```diff
@@ -0,0 +1,665 @@
+        request.sourceFrame.postUrl ||
+        homeframeUrl ||
+        "",
+    };
+
+    stackDispatch({ action: "LOAD", item: frameStackPendingItem });
+
+    const signedTransactionDataActionOrError = await signFrameAction(
+      // for transaction data we always use signer, so skip signing is false here
+      false,
+      request.signerStateActionContext
+    ).catch((e) => {
+      if (e instanceof Error) {
+        return e;
+      }
+
+      console.error(e);
+      return new Error(
+        "Unexpected error thrown by signFrameAction, see the console for details"
+      );
+    });
+
+    if (signedTransactionDataActionOrError instanceof Error) {
+      pushRequestErrorToStack({
+        endTime: new Date(),
+        frameStackPendingItem,
+        responseBody: "none",
+        responseStatus: 500,
+        startTime,
+        requestError: signedTransactionDataActionOrError,
+      });
+
+      return;
+    }
+
+    stackDispatch({
+      action: "ADD_REQUEST_DETAILS",
+      requestDetails: {
+        body: signedTransactionDataActionOrError.body,
+        searchParams: signedTransactionDataActionOrError.searchParams,
+      },
+      url: signedTransactionDataActionOrError.searchParams.get("postUrl") ?? "",
+      pendingItem: frameStackPendingItem,
+    });
+
+    signedTransactionDataActionOrError.searchParams.set(
+      "specification",
+      specification
+    );
+
+    // fetch transaction data first
+    const transactionDataProxiedUrl = `${frameActionProxy}?${signedTransactionDataActionOrError.searchParams.toString()}`;
+
+    const transactionDataResponse = await fetchRequest(
+      transactionDataProxiedUrl,
+      {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          ...extraButtonRequestPayload,
+          ...signedTransactionDataActionOrError.body,
+        }),
+      }
+    );
+    const endTime = new Date();
+
+    if (!(transactionDataResponse instanceof Response)) {
+      pushRequestErrorToStack({
+        endTime,
+        frameStackPendingItem,
+        responseBody: "none",
+        responseStatus: 500,
+        startTime,
+        requestError: transactionDataResponse,
+      });
+
+      return;
+    }
+
+    if (!transactionDataResponse.ok) {
+      await handleFailedResponse({
+        response: transactionDataResponse,
+        startTime,
+        endTime,
+        frameStackPendingItem,
+      });
+
+      return;
+    }
+
+    const transactionData =
+      (await transactionDataResponse.json()) as TransactionTargetResponse;
+
+    // get transaction id from transaction data
+    const transactionIdOrError = await onTransaction({
+      frame: request.sourceFrame,
+      frameButton: request.frameButton,
+      transactionData,
+    })
+      .then((transactionId) => {
+        if (!transactionId) {
+          return new Error("onTransaction did not return transaction id");
+        }
```

Comment:
I don't think client side errors shouldn't be presented in the same way that frame server error messages are presented. You can reproduce this by initiating the transaction but pressing cancel in the wallet
<img width="840" alt="Screenshot 2024-05-22 at 10 21 06" src="https://github.com/framesjs/frames.js/assets/5469870/b4c32545-c417-4f0c-8241-6eededeadd81">

Reply:
I don't know how the apps are handling that but in my opinion this is correct since I would expected an app to open some sort of popup where you can approve/reject the transaction and then you want the feedback to be shown but still be able to try again.

In my opinion this solution is good enough and doesn't introduce blockers for developers since they can fix the issue and try again instead of reloading.

Reply:
I am not opposed to showing feedback, I am just worried that this feedback which is unrelated to the frame server is presented in the same way as frame server errors which are an intentional UI feature

Reply:
I would recommend a toast instead

Reply:
Render package is not aware of any special error handling. I will extend the api so we can do special stuff with errors, etc.

Reply:
I added the `onError()` function that accepts an error and can be used to do whatever with it.

Also I removed one condition that is not useful anymore in FrameUI. Also request errors are now not reported as messages at all, we are leveraging 'onError()` function to do report the error, in debugger we use `toast()`